src/multiversioning.zig

```diff
@@ -0,0 +1,836 @@
+
+        self.stage = .{ .err = result };
+
+        if (self.callback) |callback| {
+            self.callback = null;
+            callback(self, result);
+        }
+    }
+
+    pub fn tick_until_ready(self: *Multiversion) !void {
+        // Either set a callback, or use this psuedo-sync interface.
+        assert(self.callback == null);
+
+        while (self.stage != .ready and self.stage != .err) {
+            try self.io.tick();
+        }
+
+        if (self.stage == .err) {
+            return self.stage.err;
+        }
+    }
+
+    pub fn exec_latest(self: *Multiversion) !noreturn {
+        log_multiversioning.info("re-executing {s}...\n", .{self.exe_path});
+
+        if (execveat(
+            self.target_fd,
+            "",
+            self.args_envp.args,
+            self.args_envp.envp,
+            os.AT.EMPTY_PATH,
+        ) == 0) {
+            unreachable;
+        } else {
+            return error.ExecveatFailed;
+        }
+    }
+
+    /// exec_release is called before a replica is fully open, and before it has transitioned to not
+    /// allocating. Therefore, standard `os.read` blocking syscalls are available.
+    /// (in any case, using blocking IO on a memfd on Linux is safe.)
+    pub fn exec_release(self: *Multiversion, release: Release) !noreturn {
+        const metadata = &self.target_metadata.?;
+
+        const index = blk: {
+            for (metadata.past.versions[0..metadata.past.count], 0..) |version, index| {
+                if (release.value == version) {
+                    break :blk index;
+                }
+            } else {
+                // This should never happen: it means that somehow ...
+                return error.VersionNotFound;
+            }
+        };
+
+        const binary_offset = metadata.past.offsets[index];
+        const binary_size = metadata.past.sizes[index];
+        const binary_checksum = metadata.past.checksums[index];
+
+        // Our target release is physically embedded in the binary. Shuffle the bytes
+        // around, so that it's at the start, then truncate the descriptor so there's nothing
+        // trailing.
+        try os.lseek_SET(self.target_fd, self.target_pack_offset.? + binary_offset);
+        const bytes_read = try os.read(self.target_fd, self.source_buffer[0..binary_size]);
+        assert(bytes_read == binary_size);
+
+        try os.lseek_SET(self.target_fd, 0);
+        const bytes_written = try os.write(self.target_fd, self.source_buffer[0..bytes_read]);
+        assert(bytes_read == bytes_written);
+
+        // Zero the remaining bytes in the file.
+        try os.ftruncate(self.target_fd, binary_size);
+
+        // Ensure the checksum matches the metadata. This could have been done above, but
+        // do it in a separate step to make sure.
+        const checksum_written = blk: {
+            try os.lseek_SET(self.target_fd, 0);
+            const bytes_read_checksum = try os.read(
```

Comment:
Is this supposed to be `bytes_read_size`? I don't understand where "checksum" comes from.

Reply:
`bytes_read_for_checksum` - it was to avoid a name clash with the outer scope's `bytes_read`.